pyron-plugin/src/main/scala/com/cloudentity/pyron/plugin/util/value/ValueResolver.scala

```diff
@@ -1,54 +1,104 @@
-    valueOrRef match {
-      case Value(value)        => value.asString
-      case BodyRef(path)       => bodyOpt.flatMap(extractBodyAttribute(_, path)).flatMap(_.asString)
-      case PathParamRef(param) => req.request.uri.pathParams.value.get(param)
-      case AuthnRef(path)      => extractAuthnCtxAttribute(req.authnCtx, path).flatMap(_.asString)
-      case HeaderRef(header, _) => req.request.headers.get(header) // always take first header value
+    (valueOrRef: @unchecked) match { // no default case, fail fast
+      case Value(value) => value.asString
+      case HostRef => Option(req.original.host).filter(_.nonEmpty)
+      case HostNameRef => resolveHostName(req)
+      case HostPortRef => resolveHostPort(req)
+      case SchemeRef => resolveScheme(req)
+      case LocalHostRef => resolveLocalHost(req)
+      case RemoteHostRef => resolveRemoteHost(req)
+      case CookieRef(cookie) => resolveCookie(req, cookie)
+      case BodyRef(path) => resolveBody(bodyOpt, path).flatMap(_.asString)
+      case PathParamRef(param) => resolvePathParam(req, param)
+      case QueryParamRef(param) => req.request.uri.query.get(param) // take first param value
+      case HeaderRef(header, AllHeaderRefType) => req.request.headers.get(header) // take first header value
+      case HeaderRef(header, FirstHeaderRefType) => req.request.headers.get(header) // take first header value
+      case AuthnRef(path) => extractAuthnCtxAttribute(req.authnCtx, path).flatMap(_.asString)
     }
 
-  def resolveString(req: RequestCtx, valueOrRef: ValueOrRef): Option[String] =
-    valueOrRef match {
-      case BodyRef(_) => resolveString(req, req.request.bodyOpt.flatMap(buf => Try(buf.toJsonObject).toOption), valueOrRef)
-      case x          => resolveString(req, None, x)
-    }
+  def resolveString(req: RequestCtx, valueOrRef: ValueOrRef): Option[String] = valueOrRef match {
+    case BodyRef(_) => resolveString(req, req.request.bodyOpt.flatMap(buf => Try(buf.toJsonObject).toOption), valueOrRef)
+    case x          => resolveString(req, None, x)
+  }
 
   def resolveListOfStrings(req: RequestCtx, bodyOpt: Option[JsonObject], valueOrRef: ValueOrRef): Option[List[String]] =
-    valueOrRef match {
+    (valueOrRef: @unchecked) match { // no default case, fail fast
       case Value(value)                          =>
         if (value.asObject.nonEmpty) circeJsonDynamicString(req, bodyOpt, value)
         else circeJsonToJsonValue(value).asListOfStrings
-      case BodyRef(path)                         => bodyOpt.flatMap(extractBodyAttribute(_, path)).flatMap(_.asListOfStrings)
-      case PathParamRef(param)                   => req.request.uri.pathParams.value.get(param).map(List(_))
-      case AuthnRef(path)                        => extractAuthnCtxAttribute(req.authnCtx, path).flatMap(_.asListOfStrings)
+      case HostRef                               => resolveHost(req).map(List(_))
```

Comment:
Please extend docs with all the new reference values https://github.com/cloudentity/pyron/blob/dev/docs/plugins/transform-request.md

Reply:
ok, will do

Reply:
done